// @ng-doc/builder/engine/entities/keywords.entity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NgDocKeywordsEntity = void 0;
const tslib_1 = require("tslib");
const path_1 = tslib_1.__importDefault(require("path"));
const rxjs_1 = require("rxjs");
const entity_1 = require("./abstractions/entity");
class NgDocKeywordsEntity extends entity_1.NgDocEntity {
    constructor() {
        super(...arguments);
        this.id = 'NgDocKeywordsEntity';
        this.isRoot = true;
        this.rootFiles = [];
        this.parent = undefined;
        this.buildCandidates = [];
    }
    build() {
        const result = this.store.getKeywords();
        return (0, rxjs_1.of)({
            result,
            entity: this,
            toBuilderOutput: (content) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                return ({
                    content: JSON.stringify(content),
                    filePath: path_1.default.join(this.context.assetsPath, 'keywords.json'),
                });
            }),
        });
    }
}
exports.NgDocKeywordsEntity = NgDocKeywordsEntity;
