// @ng-doc/app/esm2022/processors/processors/tooltip.processor.mjs
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, Input, ViewChild, } from '@angular/core';
import { NgDocTooltipDirective } from '@ng-doc/ui-kit';
import * as i0 from "@angular/core";
class NgDocTooltipWrapperComponent {
    constructor(changeDetectorRef) {
        this.changeDetectorRef = changeDetectorRef;
        this.tooltipElement = null;
    }
    ngAfterViewInit() {
        if (this.contentProjection) {
            const element = this.contentProjection.nativeElement.querySelector('[ngDocTooltip]');
            this.tooltipElement = element instanceof HTMLElement ? element : null;
            this.changeDetectorRef.detectChanges();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocTooltipWrapperComponent, deps: [{ token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "18.1.0", type: NgDocTooltipWrapperComponent, isStandalone: true, selector: "ng-doc-tooltip-wrapper", inputs: { content: "content" }, viewQueries: [{ propertyName: "contentProjection", first: true, predicate: ["contentProjection"], descendants: true, read: ElementRef, static: true }], ngImport: i0, template: `
		<div
			class="content-projection"
			[ngDocTooltip]="content ?? ''"
			[displayOrigin]="tooltipElement ?? contentProjection"
			[pointerOrigin]="tooltipElement ?? contentProjection"
			#contentProjection>
			<ng-content></ng-content>
		</div>
	`, isInline: true, styles: [".content-projection{display:unset}\n"], dependencies: [{ kind: "directive", type: NgDocTooltipDirective, selector: "[ngDocTooltip]", inputs: ["ngDocTooltip", "delay", "displayOrigin", "pointerOrigin", "positions", "canOpen", "panelClass", "minHeight", "maxHeight", "height", "minWidth", "maxWidth", "width"], outputs: ["beforeOpen", "afterOpen", "beforeClose", "afterClose"], exportAs: ["ngDocTooltip"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocTooltipWrapperComponent, decorators: [{
            type: Component,
            args: [{ selector: 'ng-doc-tooltip-wrapper', template: `
		<div
			class="content-projection"
			[ngDocTooltip]="content ?? ''"
			[displayOrigin]="tooltipElement ?? contentProjection"
			[pointerOrigin]="tooltipElement ?? contentProjection"
			#contentProjection>
			<ng-content></ng-content>
		</div>
	`, changeDetection: ChangeDetectionStrategy.OnPush, standalone: true, imports: [NgDocTooltipDirective], styles: [".content-projection{display:unset}\n"] }]
        }], ctorParameters: () => [{ type: i0.ChangeDetectorRef }], propDecorators: { content: [{
                type: Input
            }], contentProjection: [{
                type: ViewChild,
                args: ['contentProjection', { read: ElementRef, static: true }]
            }] } });
export const tooltipProcessor = {
    component: NgDocTooltipWrapperComponent,
    selector: '[ngDocTooltip]',
    extractOptions: (element) => ({
        inputs: {
            content: element.getAttribute('ngDocTooltip') ?? '',
        },
        content: [[element.cloneNode(true)]],
    }),
};
