// @ng-doc/ui-kit/esm2022/services/dialog/dialog.service.mjs
import { inject, Injectable } from '@angular/core';
import { asArray } from '@ng-doc/core';
import { NgDocOverlayContainerComponent } from '@ng-doc/ui-kit/components/overlay-container';
import { NgDocOverlayService } from '@ng-doc/ui-kit/services/overlay';
import * as i0 from "@angular/core";
export class NgDocDialogService {
    constructor() {
        this.overlayService = inject(NgDocOverlayService);
    }
    open(content, config) {
        return this.overlayService.open(content, {
            overlayContainer: NgDocOverlayContainerComponent,
            positionStrategy: this.overlayService
                .globalPositionStrategy()
                .centerHorizontally()
                .centerVertically(),
            scrollStrategy: this.overlayService.scrollStrategy().block(),
            ...config,
            panelClass: ['ng-doc-dialog', ...asArray(config?.panelClass)],
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocDialogService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocDialogService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocDialogService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
