// @ng-doc/builder/engine/actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NgDocActions = void 0;
const demo_action_1 = require("./actions/demo.action");
const demo_pane_action_1 = require("./actions/demo-pane.action");
const playground_action_1 = require("./actions/playground.action");
class NgDocActions {
    constructor(page) {
        this.page = page;
    }
    demo(className, options) {
        return this.performAction((0, demo_action_1.demoAction)(className, options)).output;
    }
    demoPane(className, options) {
        return this.performAction((0, demo_pane_action_1.demoPaneAction)(className, options)).output;
    }
    playground(playgroundId, options) {
        const output = this.performAction((0, playground_action_1.playgroundAction)(playgroundId, options));
        return output.output;
    }
    performAction(action) {
        var _a;
        const output = action(this.page.sourceFile.getProject(), this.page);
        this.page.dependencies.add(...((_a = output.dependencies) !== null && _a !== void 0 ? _a : []));
        return output;
    }
}
exports.NgDocActions = NgDocActions;
