// @ng-doc/ui-kit/fesm2022/ng-doc-ui-kit-services-notify.mjs
import { __decorate, __metadata } from 'tslib';
import * as i0 from '@angular/core';
import { Injectable } from '@angular/core';
import { notificationOpenAnimation, notificationCloseAnimation } from '@ng-doc/ui-kit/animations';
import { NgDocOverlayContainerComponent } from '@ng-doc/ui-kit/components/overlay-container';
import * as i1 from '@ng-doc/ui-kit/services/overlay';
import { NgDocOverlayService } from '@ng-doc/ui-kit/services/overlay';
import { untilDestroyed, UntilDestroy } from '@ngneat/until-destroy';
import { Subject, timer } from 'rxjs';
import { tap, switchMap } from 'rxjs/operators';

let NgDocNotifyService = class NgDocNotifyService {
    constructor(overlayService) {
        this.overlayService = overlayService;
        this.notify$ = new Subject();
        this.notify$
            .pipe(tap(() => this.overlayRef?.close()), tap((content) => this.openOverlay(content)), switchMap(() => timer(2000)), untilDestroyed(this))
            .subscribe(() => this.overlayRef?.close());
    }
    notify(content) {
        this.notify$.next(content);
    }
    openOverlay(content) {
        this.overlayRef = this.overlayService.open(content, {
            overlayContainer: NgDocOverlayContainerComponent,
            panelClass: 'ng-doc-notify',
            positionStrategy: this.overlayService
                .globalPositionStrategy()
                .bottom('10px')
                .centerHorizontally(),
            openAnimation: notificationOpenAnimation,
            closeAnimation: notificationCloseAnimation,
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocNotifyService, deps: [{ token: i1.NgDocOverlayService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocNotifyService, providedIn: 'root' }); }
};
NgDocNotifyService = __decorate([
    UntilDestroy(),
    __metadata("design:paramtypes", [NgDocOverlayService])
], NgDocNotifyService);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocNotifyService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: () => [{ type: i1.NgDocOverlayService }] });

/**
 * Generated bundle index. Do not edit.
 */

export { NgDocNotifyService };
