// @ng-doc/ui-kit/fesm2022/ng-doc-ui-kit-components-floated-content.mjs
import { __decorate, __metadata } from 'tslib';
import * as i0 from '@angular/core';
import { Component, ChangeDetectionStrategy, Input, HostBinding } from '@angular/core';
import { ngDocMakePure } from '@ng-doc/ui-kit/decorators';
import { toElement } from '@ng-doc/ui-kit/helpers';

class NgDocFloatedContentComponent {
    constructor(elementRef, renderer, ngZone) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.ngZone = ngZone;
        this.propertyName = '';
        this.alignTo = 'left';
    }
    ngAfterViewChecked() {
        if (this.bindTo && this.alignTo) {
            this.ngZone.runOutsideAngular(() => Promise.resolve().then(() => this.setPadding(this.elementRef.nativeElement.offsetWidth)));
        }
    }
    setPadding(padding) {
        if (this.bindTo && this.propertyName) {
            this.renderer.setStyle(toElement(this.bindTo), this.propertyName, padding ? `${padding}px` : null, 2);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocFloatedContentComponent, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "18.1.0", type: NgDocFloatedContentComponent, isStandalone: true, selector: "[ng-doc-floated-content]", inputs: { bindTo: "bindTo", propertyName: "propertyName", alignTo: "alignTo" }, host: { properties: { "attr.data-ng-doc-align": "this.alignTo" } }, ngImport: i0, template: ` <ng-content></ng-content> `, isInline: true, styles: [":host{position:absolute;display:flex;top:0;align-items:center;height:100%;pointer-events:none}:host[data-ng-doc-align=left]{left:0}:host[data-ng-doc-align=right]{right:0}:host ::ng-deep>*{pointer-events:auto}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
__decorate([
    ngDocMakePure,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Number]),
    __metadata("design:returntype", void 0)
], NgDocFloatedContentComponent.prototype, "setPadding", null);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocFloatedContentComponent, decorators: [{
            type: Component,
            args: [{ selector: '[ng-doc-floated-content]', template: ` <ng-content></ng-content> `, changeDetection: ChangeDetectionStrategy.OnPush, standalone: true, styles: [":host{position:absolute;display:flex;top:0;align-items:center;height:100%;pointer-events:none}:host[data-ng-doc-align=left]{left:0}:host[data-ng-doc-align=right]{right:0}:host ::ng-deep>*{pointer-events:auto}\n"] }]
        }], ctorParameters: () => [{ type: i0.ElementRef }, { type: i0.Renderer2 }, { type: i0.NgZone }], propDecorators: { bindTo: [{
                type: Input
            }], propertyName: [{
                type: Input
            }], alignTo: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['attr.data-ng-doc-align']
            }], setPadding: [] } });

/**
 * Generated bundle index. Do not edit.
 */

export { NgDocFloatedContentComponent };
