// @ng-doc/ui-kit/fesm2022/ng-doc-ui-kit-directives-data.mjs
import * as i0 from '@angular/core';
import { Directive } from '@angular/core';

class NgDocDataDirective {
    constructor(template, viewContainerRef) {
        this.template = template;
        this.viewContainerRef = viewContainerRef;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocDataDirective, deps: [{ token: i0.TemplateRef }, { token: i0.ViewContainerRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.1.0", type: NgDocDataDirective, isStandalone: true, selector: "[ngDocData]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocDataDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngDocData]',
                    standalone: true,
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef }, { type: i0.ViewContainerRef }] });

/**
 * Generated bundle index. Do not edit.
 */

export { NgDocDataDirective };
