// @ng-doc/app/esm2022/helpers/generate-toc.mjs
import { asArray } from '@ng-doc/core/helpers/as-array';
/**
 * Generate table of contents, only for headings with id
 * @param container
 * @param headings
 */
export function generateToc(container) {
    const headings = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'];
    const headingElements = Array.from(container.querySelectorAll(headings.join(', '))).filter((heading) => heading.id);
    const levels = asArray(new Set(headingElements.map(levelFromTagName).sort()));
    return headingElements.reduce((map, heading) => {
        const headingLevel = levelFromTagName(heading);
        const anchor = heading.querySelector('a.ng-doc-header-link');
        if (anchor) {
            map.push({
                title: heading.textContent?.trim() ?? '',
                element: heading,
                path: anchor.pathname,
                hash: anchor.hash.replace('#', ''),
                level: levels.indexOf(headingLevel) + 1,
            });
        }
        return map;
    }, []);
}
/**
 *
 * @param heading
 */
function levelFromTagName(heading) {
    return Number(heading.tagName.toLowerCase().replace(/[a-z]*/g, '') || 1);
}
