// @ng-doc/ui-kit/esm2022/components/selection/selection-host.directive.mjs
import { Directive } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { distinctUntilChanged } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class NgDocSelectionHostDirective {
    constructor() {
        this.origins = new Set();
        this.selectedChange = new ReplaySubject();
    }
    get selectedChange$() {
        return this.selectedChange.pipe(distinctUntilChanged());
    }
    addOrigin(origin) {
        this.origins.add(origin);
    }
    removeOrigin(origin) {
        this.origins.delete(origin);
        if (this.selected === origin) {
            this.changeSelected(origin, false);
        }
    }
    changeSelected(origin, selected) {
        this.selected =
            this.selected === origin || selected ? (selected ? origin : undefined) : this.selected;
        this.selectedChange.next(this.selected?.elementRef?.nativeElement ?? undefined);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocSelectionHostDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.1.0", type: NgDocSelectionHostDirective, isStandalone: true, selector: "[ngDocSelectionHost]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocSelectionHostDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngDocSelectionHost]',
                    standalone: true,
                }]
        }] });
