// @ng-doc/builder/helpers/angular/get-component-decorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getComponentDecorator = getComponentDecorator;
const ts_morph_1 = require("ts-morph");
const string_expression_1 = require("../string-expression");
const typescript_1 = require("../typescript");
/**
 *    Resolves the component decorator and return its properties.
 * @param cls - class declaration
 */
function getComponentDecorator(cls) {
    const decorator = cls.getDecorator('Component');
    const decoratorArgument = decorator === null || decorator === void 0 ? void 0 : decorator.getArguments()[0];
    if (ts_morph_1.Node.isObjectLiteralExpression(decoratorArgument)) {
        const standaloneProperty = decoratorArgument.getProperty('standalone');
        const selectorProperty = decoratorArgument.getProperty('selector');
        const templateProperty = decoratorArgument.getProperty('templateUrl');
        const styleUrlProperty = decoratorArgument.getProperty('styleUrl');
        const styleUrlsProperty = decoratorArgument.getProperty('styleUrls');
        return {
            standalone: (0, string_expression_1.stringExpression)((0, typescript_1.getPropertyAssignment)(standaloneProperty)),
            selector: (0, string_expression_1.stringExpression)((0, typescript_1.getPropertyAssignment)(selectorProperty)),
            templateUrl: (0, string_expression_1.stringExpression)((0, typescript_1.getPropertyAssignment)(templateProperty)),
            styleUrl: (0, string_expression_1.stringExpression)((0, typescript_1.getPropertyAssignment)(styleUrlProperty)),
            styleUrls: (0, string_expression_1.stringExpression)((0, typescript_1.getPropertyAssignment)(styleUrlsProperty)),
        };
    }
    return undefined;
}
