// @ng-doc/ui-kit/esm2022/animations/overlay.mjs
import { animate, style } from '@angular/animations';
export const dropdownOpenAnimation = [
    style({ transform: 'scale(0.9)', opacity: 0 }),
    animate('120ms cubic-bezier(0.25, 0.8, 0.25, 1)', style({ transform: 'scale(1)', opacity: 1 })),
];
export const dropdownCloseAnimation = [
    style({ transform: 'scaleY(1)', opacity: 1 }),
    animate('120ms cubic-bezier(0.25, 0.8, 0.25, 1)', style({ transform: 'scaleY(0.9)', opacity: 0 })),
];
export const tooltipOpenAnimation = [
    style({ transform: 'scale(0.8)', opacity: 0 }),
    animate('120ms cubic-bezier(0.25, 0.8, 0.25, 1)', style({ transform: 'scale(1)', opacity: 1 })),
];
export const tooltipCloseAnimation = [
    style({ transform: 'scale(1)', opacity: 1 }),
    animate('120ms cubic-bezier(0.25, 0.8, 0.25, 1)', style({ transform: 'scale(0.8)', opacity: 0 })),
];
export const notificationOpenAnimation = [
    style({ transform: 'scale(0.7)', opacity: 0 }),
    animate('120ms cubic-bezier(0.25, 0.8, 0.25, 1)', style({ transform: 'scale(1)', opacity: 1 })),
];
export const notificationCloseAnimation = [
    style({ transform: 'scale(1)', opacity: 1 }),
    animate('120ms cubic-bezier(0.25, 0.8, 0.25, 1)', style({ transform: 'scale(0.7)', opacity: 0 })),
];
