// di-controls/fesm2022/di-controls-directives-control-silencer.mjs
import * as i0 from '@angular/core';
import { Directive } from '@angular/core';
import { provideHostControl } from 'di-controls/tokens';
import { DIControl } from 'di-controls/controls';

class DIControlSilencerDirective extends DIControl {
    constructor() {
        super();
    }
    registerControl() {
        // do nothing
    }
    unregisterControl() {
        // do nothing
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: DIControlSilencerDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.0.0", type: DIControlSilencerDirective, isStandalone: true, selector: "[diControlSilencer]", providers: [provideHostControl(DIControlSilencerDirective)], usesInheritance: true, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: DIControlSilencerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[diControlSilencer]',
                    standalone: true,
                    providers: [provideHostControl(DIControlSilencerDirective)],
                }]
        }], ctorParameters: function () { return []; } });

/**
 * Generated bundle index. Do not edit.
 */

export { DIControlSilencerDirective };
