// @ng-doc/ui-kit/esm2022/directives/dropdown-origin/dropdown-origin.directive.mjs
import { CdkOverlayOrigin } from '@angular/cdk/overlay';
import { Directive, ElementRef } from '@angular/core';
import { NgDocOverlayHost } from '@ng-doc/ui-kit/classes/overlay-host';
import * as i0 from "@angular/core";
export class NgDocDropdownOriginDirective extends CdkOverlayOrigin {
    constructor(origin) {
        super(origin);
        this.origin = origin;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocDropdownOriginDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.1.0", type: NgDocDropdownOriginDirective, isStandalone: true, selector: "[ngDocDropdownOrigin]", providers: [
            {
                provide: NgDocOverlayHost,
                useExisting: NgDocDropdownOriginDirective,
            },
        ], exportAs: ["ngDocDropdownOrigin"], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocDropdownOriginDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngDocDropdownOrigin]',
                    exportAs: 'ngDocDropdownOrigin',
                    providers: [
                        {
                            provide: NgDocOverlayHost,
                            useExisting: NgDocDropdownOriginDirective,
                        },
                    ],
                    standalone: true,
                }]
        }], ctorParameters: () => [{ type: i0.ElementRef }] });
