// @ng-doc/builder/engine/entities/context.entity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NgDocContextEntity = void 0;
const tslib_1 = require("tslib");
const path_1 = tslib_1.__importDefault(require("path"));
const rxjs_1 = require("rxjs");
const nunjucks_1 = require("../nunjucks");
const entity_1 = require("./abstractions/entity");
class NgDocContextEntity extends entity_1.NgDocEntity {
    constructor() {
        super(...arguments);
        this.id = 'NgDocContextEntity';
        this.isRoot = true;
        this.rootFiles = [];
        this.parent = undefined;
        this.buildCandidates = [];
    }
    build() {
        const entities = this.store.getRootEntitiesForBuild();
        const result = (0, nunjucks_1.renderTemplate)('./context.ts.nunj', { context: { entities } });
        return (0, rxjs_1.of)({
            result,
            entity: this,
            toBuilderOutput: (content) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                return ({
                    content,
                    filePath: path_1.default.join(this.context.buildPath, 'ng-doc.context.ts'),
                });
            }),
        });
    }
}
exports.NgDocContextEntity = NgDocContextEntity;
