// @ng-doc/builder/engine/entities/index-file.entity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NgDocIndexFileEntity = void 0;
const tslib_1 = require("tslib");
const path = tslib_1.__importStar(require("path"));
const rxjs_1 = require("rxjs");
const nunjucks_1 = require("../nunjucks");
const entity_1 = require("./abstractions/entity");
const cache_1 = require("./cache");
let NgDocIndexFileEntity = (() => {
    let _classDecorators = [(0, cache_1.CachedEntity)()];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _classSuper = entity_1.NgDocEntity;
    let _instanceExtraInitializers = [];
    let _get_outputPath_decorators;
    var NgDocIndexFileEntity = _classThis = class extends _classSuper {
        constructor() {
            super(...arguments);
            this.id = (tslib_1.__runInitializers(this, _instanceExtraInitializers), `NgDocIndexFile`);
            this.isRoot = false;
            this.parent = undefined;
            this.rootFiles = [];
            this.buildCandidates = [];
        }
        get outputPath() {
            return path.join(this.context.buildPath, 'index.ts');
        }
        build() {
            return (0, rxjs_1.of)({
                result: (0, nunjucks_1.renderTemplate)('./index.ts.nunj'),
                entity: this,
                toBuilderOutput: (content) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                    return ({
                        content,
                        filePath: this.outputPath,
                    });
                }),
            });
        }
    };
    tslib_1.__setFunctionName(_classThis, "NgDocIndexFileEntity");
    (() => {
        var _a;
        const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create((_a = _classSuper[Symbol.metadata]) !== null && _a !== void 0 ? _a : null) : void 0;
        _get_outputPath_decorators = [(0, cache_1.CachedFilesGetter)()];
        tslib_1.__esDecorate(_classThis, null, _get_outputPath_decorators, { kind: "getter", name: "outputPath", static: false, private: false, access: { has: obj => "outputPath" in obj, get: obj => obj.outputPath }, metadata: _metadata }, null, _instanceExtraInitializers);
        tslib_1.__esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
        NgDocIndexFileEntity = _classThis = _classDescriptor.value;
        if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        tslib_1.__runInitializers(_classThis, _classExtraInitializers);
    })();
    return NgDocIndexFileEntity = _classThis;
})();
exports.NgDocIndexFileEntity = NgDocIndexFileEntity;
