// @ng-doc/builder/engine/entities/indexes.entity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NgDocIndexesEntity = void 0;
const tslib_1 = require("tslib");
const path_1 = tslib_1.__importDefault(require("path"));
const rxjs_1 = require("rxjs");
const entity_1 = require("./abstractions/entity");
class NgDocIndexesEntity extends entity_1.NgDocEntity {
    constructor() {
        super(...arguments);
        this.id = 'NgDocIndexesEntity';
        this.isRoot = true;
        this.rootFiles = [];
        this.parent = undefined;
        this.buildCandidates = [];
    }
    build() {
        const result = this.store
            .asArray()
            .map((entity) => entity.indexes)
            .flat();
        return (0, rxjs_1.of)({
            result,
            entity: this,
            toBuilderOutput: (content) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                return ({
                    content: JSON.stringify(content, null, 2),
                    filePath: path_1.default.join(this.context.assetsPath, 'indexes.json'),
                });
            }),
        });
    }
}
exports.NgDocIndexesEntity = NgDocIndexesEntity;
