// @ng-doc/builder/engine/entities/routes.entity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NgDocRoutesEntity = void 0;
const tslib_1 = require("tslib");
const path = tslib_1.__importStar(require("path"));
const rxjs_1 = require("rxjs");
const nunjucks_1 = require("../nunjucks");
const entity_1 = require("./abstractions/entity");
class NgDocRoutesEntity extends entity_1.NgDocEntity {
    constructor() {
        super(...arguments);
        this.id = `NgDocRoutes`;
        this.isRoot = false;
        this.parent = undefined;
        this.rootFiles = [];
        this.buildCandidates = [];
    }
    build() {
        const entities = this.store.getRootEntitiesForBuild();
        const outFolder = this.context.buildPath;
        const result = (0, nunjucks_1.renderTemplate)('./routing.ts.nunj', {
            context: { entities, outFolder },
        });
        return (0, rxjs_1.of)({
            result,
            entity: this,
            toBuilderOutput: (content) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                return ({
                    content,
                    filePath: path.join(outFolder, 'ng-doc.routing.ts'),
                });
            }),
        });
    }
}
exports.NgDocRoutesEntity = NgDocRoutesEntity;
