// @ng-doc/ui-kit/esm2022/components/expander/expander.component.mjs
import { NgIf } from '@angular/common';
import { ChangeDetectionStrategy, Component, HostBinding, Input } from '@angular/core';
import { expandCollapseAnimation, preventInitialChildAnimations } from '@ng-doc/ui-kit/animations';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import * as i0 from "@angular/core";
import * as i1 from "@tinkoff/ng-polymorpheus";
/** Component helps to expand or collapse content */
export class NgDocExpanderComponent {
    constructor() {
        /** Change expand state */
        this.expanded = false;
        /** Expander content */
        this.content = '';
        /** Closed height could be used to show preview of the content */
        this.from = 0;
    }
    toggle() {
        this.expanded = !this.expanded;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocExpanderComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "18.1.0", type: NgDocExpanderComponent, isStandalone: true, selector: "ng-doc-expander", inputs: { expanded: "expanded", content: "content", from: "from" }, host: { properties: { "@preventInitialChild": "this.preventInitialChild" } }, ngImport: i0, template: "<div\n\t*ngIf=\"expanded || from\"\n\t[@expandCollapse]=\"{ value: expanded, params: { from: from + 'px', opacity: from ? 1 : 0 } }\">\n\t<ng-container *polymorpheusOutlet=\"content as text\">{{ text }}</ng-container>\n</div>\n", styles: [":host{display:block}\n"], dependencies: [{ kind: "directive", type: NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "ngmodule", type: PolymorpheusModule }, { kind: "directive", type: i1.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }], animations: [preventInitialChildAnimations, expandCollapseAnimation], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocExpanderComponent, decorators: [{
            type: Component,
            args: [{ animations: [preventInitialChildAnimations, expandCollapseAnimation], selector: 'ng-doc-expander', changeDetection: ChangeDetectionStrategy.OnPush, standalone: true, imports: [NgIf, PolymorpheusModule], template: "<div\n\t*ngIf=\"expanded || from\"\n\t[@expandCollapse]=\"{ value: expanded, params: { from: from + 'px', opacity: from ? 1 : 0 } }\">\n\t<ng-container *polymorpheusOutlet=\"content as text\">{{ text }}</ng-container>\n</div>\n", styles: [":host{display:block}\n"] }]
        }], propDecorators: { expanded: [{
                type: Input
            }], content: [{
                type: Input
            }], from: [{
                type: Input
            }], preventInitialChild: [{
                type: HostBinding,
                args: ['@preventInitialChild']
            }] } });
