// @ng-doc/app/fesm2022/ng-doc-app-constants.mjs
const NG_DOC_STORE_THEME_KEY = 'ng-doc-theme-id';

/**
 * Dark theme. You can set it manually by calling `NgDocThemeService.set` method.
 *
 * ```typescript
 * import {Component} from '@angular/core';
 * import {NgDocThemeService, NG_DOC_NIGHT_THEME} from '@ng-doc/app';
 *
 * @Component()
 * export class AppComponent {
 *    constructor(
 *        private readonly themeService: NgDocThemeService,
 *    ) {
 *        this.themeService.setTheme(NG_DOC_NIGHT_THEME.id);
 *    }
 * }
 * ```
 *
 * Or you can set it as default theme in your app module:
 * ```typescript
 * import {NgModule} from '@angular/core';
 * import {NgDocModule, NG_DOC_NIGHT_THEME} from '@ng-doc/app';
 * import {AppComponent} from './app.component';
 *
 * @NgModule({
 *    declarations: [AppComponent],
 *    imports: [NgDocModule.forRoot({defaultTheme: NG_DOC_NIGHT_THEME.id})],
 *    bootstrap: [AppComponent],
 * })
 * export class AppModule {}
 * ```
 */
const NG_DOC_NIGHT_THEME = {
    id: 'ng-doc-night',
    path: 'assets/ng-doc/app/themes/ng-doc-night.css',
};

/**
 * Generated bundle index. Do not edit.
 */

export { NG_DOC_NIGHT_THEME, NG_DOC_STORE_THEME_KEY };
