// @ng-doc/app/fesm2022/ng-doc-app-pipes-filter-by-text.mjs
import * as i0 from '@angular/core';
import { Pipe } from '@angular/core';

class NgDocFilterByTextPipe {
    transform(items, searchTerm, matcher = (item, s) => String(item) === s) {
        return items?.filter((item) => matcher(item, searchTerm)) ?? [];
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocFilterByTextPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.1.0", ngImport: i0, type: NgDocFilterByTextPipe, isStandalone: true, name: "ngDocFilterByText" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocFilterByTextPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'ngDocFilterByText',
                    standalone: true,
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { NgDocFilterByTextPipe };
