// @ng-doc/builder/classes/observable-set.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObservableSet = void 0;
const core_1 = require("@ng-doc/core");
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
class ObservableSet {
    constructor(values) {
        this.collection = new Set();
        this.changes$ = new rxjs_1.ReplaySubject();
        values && this.add(...values);
    }
    asArray() {
        return (0, core_1.asArray)(this.collection);
    }
    changes() {
        return this.changes$.pipe((0, operators_1.map)(() => (0, core_1.asArray)(this.collection)));
    }
    add(...values) {
        if (values.some((value) => !this.collection.has(value))) {
            values.forEach((value) => this.collection.add(value));
            values.length && this.changes$.next();
        }
    }
    fill(...values) {
        this.collection.clear();
        this.add(...values);
    }
    delete(value) {
        this.collection.delete(value);
        this.changes$.next();
    }
    clear() {
        this.collection.clear();
        this.changes$.next();
    }
}
exports.ObservableSet = ObservableSet;
