// @ng-doc/builder/engine/entities/abstractions/module.entity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NgDocModuleEntity = void 0;
const tslib_1 = require("tslib");
const path = tslib_1.__importStar(require("path"));
const helpers_1 = require("../../../helpers");
const cache_1 = require("../cache");
const file_entity_1 = require("./file.entity");
let NgDocModuleEntity = (() => {
    var _a;
    let _classSuper = file_entity_1.NgDocFileEntity;
    let _instanceExtraInitializers = [];
    let _get_modulePath_decorators;
    return _a = class NgDocModuleEntity extends _classSuper {
            constructor() {
                super(...arguments);
                /**
                 * Title of the module file.
                 */
                this.moduleFileName = (tslib_1.__runInitializers(this, _instanceExtraInitializers), 'module.ts');
            }
            get folderPath() {
                var _b, _c;
                return path.join((_c = (_b = this.parent) === null || _b === void 0 ? void 0 : _b.folderPath) !== null && _c !== void 0 ? _c : this.context.guidesPath, this.folderName);
            }
            /**
             * Returns paths to the module in generated folder
             * @type {string}
             */
            get modulePath() {
                return path.join(this.folderPath, this.moduleFileName);
            }
            /**
             * Returns relative paths to the module in generated folder that could be used for import
             * @type {string}
             */
            get moduleImport() {
                return (0, helpers_1.posix)(path.relative(this.context.context.workspaceRoot, this.modulePath)).replace(/.ts$/, '');
            }
            destroy() {
                super.destroy();
            }
            removeArtifacts() {
                super.removeArtifacts();
            }
        },
        (() => {
            var _b;
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create((_b = _classSuper[Symbol.metadata]) !== null && _b !== void 0 ? _b : null) : void 0;
            _get_modulePath_decorators = [(0, cache_1.CachedFilesGetter)()];
            tslib_1.__esDecorate(_a, null, _get_modulePath_decorators, { kind: "getter", name: "modulePath", static: false, private: false, access: { has: obj => "modulePath" in obj, get: obj => obj.modulePath }, metadata: _metadata }, null, _instanceExtraInitializers);
            if (_metadata) Object.defineProperty(_a, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
        })(),
        _a;
})();
exports.NgDocModuleEntity = NgDocModuleEntity;
