// di-controls/esm2022/controls/collection-control.mjs
import { Directive, Input } from '@angular/core';
import { SetCompare } from 'di-controls/classes';
import { DI_DEFAULT_COMPARE } from 'di-controls/constants';
import { DIControl } from './control';
import { DIStateControl } from './state-control';
import * as i0 from "@angular/core";
/**
 * `DICollectionControl` can be used to implement array controls (checkbox group, radio group, chips, etc.).
 * It has an additional integration with `DIStateControl` that allows you to use it as a host for
 * `DIStateControl` controls. If you use `DIStateControl` as a child control, then `DICollectionControl`
 * will update its model when the child control is checked or unchecked, so `DICollectionControl` will
 * contain only checked values.
 *
 * It also works with other controls, but their model should be an array.
 *
 * > **Warning**
 * > If child control model is updated with non-array value, then `DICollectionControl` will be updated with `null`.
 *
 * ## Creating a control
 * To create a control you need to extend your `@Component` or `@Directive` from `DICollectionControl` class.
 * After that your control will be able to work with `NgModel`, `FormControl`.
 *
 * ```ts fileName="custom-control.component.ts"
 * @Component({})
 * export class CustomControlComponent extends DICollectionControl<string> {
 *   constructor() {
 *    super();
 *  }
 * }
 *  ```
 *
 * ## Registering as a host
 * By default your control can work only with `NgModel` and `FormControl`. But you can register your control as a host
 * for another controls, then your control will be able to update them and accept updates from them. To do that you need to
 * use `provideHostControl` function.
 *
 * ```ts {2} fileName="custom-control.component.ts"
 * @Component({
 *   providers: [provideHostControl(CustomControlComponent)],
 * })
 * export class CustomControlComponent extends DICollectionControl<string> {
 *   constructor() {
 *     super();
 *   }
 * }
 * ```
 *
 * ## Injecting host control
 * By default your control doesn't communicate with host controls. But you can inject host control and put it
 * into `super` call. This will register your control in the host control and start communication between them.
 *
 * > **Note**
 * > If you register your control as a host for another controls, then you can inject it
 * > only with `skipSelf` option.
 *
 * ```ts {5} fileName="custom-control.component.ts"
 * @Component({})
 * export class CustomControlComponent extends DICollectionControl<string> {
 *   constructor() {
 *     // we add `optional` option to make it possible to use this control without host
 *     super({host: injectHostControl({optional: true})});
 *   }
 * }
 * ```
 *
 * ## Getting model
 * To get model you need to use `model` property. It will return model for the current control.
 *
 * ```ts {9} fileName="custom-control.component.ts"
 * @Component({})
 * export class CustomControlComponent extends DICollectionControl<string> {
 *   constructor() {
 *     super();
 *   }
 *
 *   @HostListener('click')
 *   onClick() {
 *     console.log(this.model());
 *   }
 * }
 * ```
 *
 * ## Updating model
 * To update model you need to call `updateModel` method. It will update model for the current control and all
 * children controls, as well as for the `NgModel` or `FormControl`.
 *
 * ```ts {9} fileName="custom-control.component.ts"
 * @Component({})
 * export class CustomControlComponent extends DICollectionControl<string> {
 *   constructor() {
 *     super();
 *   }
 *
 *   @HostListener('click')
 *   onClick() {
 *     this.updateModel(['new value']);
 *   }
 * }
 * ```
 * ## Catching updates
 * Sometimes you may need to catch updates from different sources. For example, to update the value of the native
 * input element. To do this, you can provide the `onIncomingUpdate` hook.
 *
 * ```ts {6} fileName="custom-control.component.ts"
 * @Component({})
 * export class CustomControlComponent extends DICollectionControl<string> {
 *   constructor() {
 *     super({
 *       onIncomingUpdate: (value: string[] | null) => {
 *         this.elementRef.nativeElement.value = value;
 *       },
 *     });
 *   }
 * }
 * ```
 *
 * ## Comparing values
 * By default `DICollectionControl` uses `===` operator to compare values. It may be not enough for some cases,
 * for example, when you want to compare immutable objects. To solve this problem you can provide `compareFn`
 * function and provide your control as a `DICompareHost`. It will be used to compare values in the array.
 *
 * > **Warning**
 * > This function will be used by `DIStateControl` to compare values, to set checked state.
 * > Don't forget to inject `DICompareHost` in your `DIStateControl` to make it work.
 *
 * ```ts {2} fileName="custom-control.component.ts"
 * @Component({
 *   providers: [provideCompareHost(CustomControlComponent)],
 * })
 * export class CustomControlComponent extends DICollectionControl<string> {
 *  constructor() {
 *    super()
 *  }
 * }
 * ```
 */
class DICollectionControl extends DIControl {
    config;
    /**
     * Function that will be used to compare values in the array.
     * Useful when you want to compare objects by some property (e.g. `id`).
     */
    compareFn = DI_DEFAULT_COMPARE;
    proxyModel = new SetCompare();
    constructor(config) {
        super(config);
        this.config = config;
    }
    updateModel(obj) {
        this.proxyModel = new SetCompare(this.compareFn, obj);
        super.updateModel(obj);
    }
    writeValue(value) {
        this.proxyModel = new SetCompare(this.compareFn, value);
        super.writeValue(value);
    }
    writeValueFromHost(obj) {
        this.proxyModel = new SetCompare(this.compareFn, obj);
        super.writeValueFromHost(obj);
    }
    childControlChange(control, value) {
        this.updateFrom = control;
        this.updateModel(this.getNewModel(control, value));
        this.incomingUpdate && this.incomingUpdate(this.model());
    }
    getNewModel(control, updates) {
        if (control instanceof DIStateControl) {
            control.checked()
                ? this.proxyModel.add(control.value)
                : this.proxyModel.delete(control.value);
        }
        else if (Array.isArray(updates)) {
            this.proxyModel = new SetCompare(this.compareFn, updates);
        }
        else {
            this.proxyModel = new SetCompare(this.compareFn);
            return null;
        }
        return this.proxyModel.toArray();
    }
    updateControl(control) {
        if (control instanceof DIStateControl) {
            control.writeValueFromHost(this.proxyModel.has(control.value) ? control.value : false);
        }
        else {
            control.writeValueFromHost(this.model());
        }
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: DICollectionControl, deps: "invalid", target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.0.0", type: DICollectionControl, inputs: { compareFn: "compareFn" }, usesInheritance: true, ngImport: i0 });
}
export { DICollectionControl };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: DICollectionControl, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: undefined }]; }, propDecorators: { compareFn: [{
                type: Input
            }] } });
