// di-controls/esm2022/controls/control-value-accessor.mjs
import { ChangeDetectorRef, Directive, ElementRef, inject, Input, Renderer2, signal } from '@angular/core';
import { NgControl } from '@angular/forms';
import { EMPTY_FUNCTION } from 'di-controls/constants';
import { hasValue } from 'di-controls/helpers';
import * as i0 from "@angular/core";
/**
 * Base implementation of ControlValueAccessor
 */
class DIControlValueAccessor {
    incomingUpdate;
    model = signal(null);
    ngControl;
    elementRef = inject(ElementRef);
    changeDetectorRef;
    renderer = inject(Renderer2);
    disabledValue = signal(false);
    touch = EMPTY_FUNCTION;
    change = EMPTY_FUNCTION;
    constructor(incomingUpdate) {
        this.incomingUpdate = incomingUpdate;
        this.ngControl = inject(NgControl, { optional: true, self: true });
        this.changeDetectorRef = inject(ChangeDetectorRef);
        if (this.ngControl) {
            this.ngControl.valueAccessor = this;
        }
    }
    /**
     * Returns true if the control is not empty.
     */
    get hasValue() {
        return hasValue(this.model());
    }
    /**
     * Returns true if the control is disabled.
     */
    get disabled() {
        return this.disabledValue();
    }
    set disabled(isDisabled) {
        this.setDisabledState(isDisabled);
    }
    /**
     * Method is called by the forms API.
     *
     * @param fn - callback function to register on value change
     * @internal
     */
    registerOnChange(fn) {
        this.change = fn;
    }
    /**
     * Method is called by the forms API.
     *
     * @param fn - callback function to register on touch
     * @internal
     */
    registerOnTouched(fn) {
        this.touch = fn;
    }
    /**
     * Method is called by the forms API to write to the view when programmatic changes from model to view are requested.
     *
     * @param obj - new value
     * @internal
     */
    writeValue(obj) {
        if (this.model() !== obj) {
            this.update(obj);
        }
    }
    /**
     * Updates the model. Use this method to update model from your Control
     *
     * @param value - new value
     */
    updateModel(value) {
        this.model.set(value);
        this.change(this.model());
        this.changeDetectorRef.markForCheck();
    }
    /**
     * Method is called by the forms API to write to the view when programmatic changes from model to view are requested.
     *
     * @param isDisabled - new value
     * @internal
     */
    setDisabledState(isDisabled) {
        this.disabledValue.set(isDisabled);
        this.disabledValue()
            ? this.renderer.setAttribute(this.elementRef.nativeElement, 'disabled', 'true')
            : this.renderer.removeAttribute(this.elementRef.nativeElement, 'disabled');
        this.renderer.setAttribute(this.elementRef.nativeElement, 'aria-disabled', `${this.disabledValue()}`);
    }
    update(value) {
        this.model.set(value);
        this.incomingUpdate && this.incomingUpdate(value);
        this.changeDetectorRef.markForCheck();
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: DIControlValueAccessor, deps: "invalid", target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.0.0", type: DIControlValueAccessor, inputs: { disabled: "disabled" }, ngImport: i0 });
}
export { DIControlValueAccessor };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.0", ngImport: i0, type: DIControlValueAccessor, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: undefined }]; }, propDecorators: { disabled: [{
                type: Input
            }] } });
