// @ng-doc/app/esm2022/tokens/page-processor.mjs
import { InjectionToken } from '@angular/core';
import { asArray } from '@ng-doc/core';
export const NG_DOC_PAGE_PROCESSOR = new InjectionToken('NG_DOC_PAGE_PROCESSOR');
export const NG_DOC_PAGE_CUSTOM_PROCESSOR = new InjectionToken('NG_DOC_PAGE_CUSTOM_PROCESSOR');
/**
 * Provide a main processor to replace html nodes with an Angular component.
 * Main processors are run before custom processors.
 * @param processors - The processor to provide.
 */
export function provideMainPageProcessor(processors) {
    return asArray(processors).map((processor) => ({
        provide: NG_DOC_PAGE_PROCESSOR,
        useValue: processor,
        multi: true,
    }));
}
/**
 * Provide a processor to replace html nodes with an Angular component.
 * @param processors - The processor to provide.
 */
export function providePageProcessor(processors) {
    return asArray(processors).map((processor) => ({
        provide: NG_DOC_PAGE_CUSTOM_PROCESSOR,
        useValue: processor,
        multi: true,
    }));
}
