// @ng-doc/app/fesm2022/ng-doc-app-pipes-decode-uri-component.mjs
import * as i0 from '@angular/core';
import { Pipe } from '@angular/core';

class NgDocDecodeUriComponentPipe {
    transform(value) {
        try {
            return decodeURIComponent(value);
        }
        catch {
            return value;
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocDecodeUriComponentPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.1.0", ngImport: i0, type: NgDocDecodeUriComponentPipe, isStandalone: true, name: "decodeUriComponent" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocDecodeUriComponentPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'decodeUriComponent',
                    standalone: true,
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { NgDocDecodeUriComponentPipe };
