// @ng-doc/app/esm2022/processors/processors/tabs.processor.mjs
import { Renderer2 } from '@angular/core';
import { NgDocTabsComponent } from '@ng-doc/app/components/tabs';
export const tabsProcessor = {
    component: NgDocTabsComponent,
    selector: 'ng-doc-tab',
    nodeToReplace: (element, injector) => {
        const renderer = injector.get(Renderer2);
        const anchor = renderer.createElement('div');
        return element.parentNode?.insertBefore(anchor, element) ?? element;
    },
    extractOptions: (element, root) => {
        const group = element.getAttribute('group') ?? '';
        const tabs = Array.from(root.querySelectorAll(`ng-doc-tab[group="${group}"]`));
        const mappedTabs = tabs.map((tab) => ({
            title: tab.getAttribute('name') ?? '',
            content: tab,
            icon: tab.getAttribute('icon') || undefined,
            active: tab.hasAttribute('active'),
        }));
        tabs.forEach((tab) => tab.remove());
        return {
            inputs: {
                tabs: mappedTabs,
            },
        };
    },
};
