// @ng-doc/ui-kit/esm2022/components/selection/selection-origin.directive.mjs
import { Directive, ElementRef, Input } from '@angular/core';
import { NgDocSelectionHostDirective } from './selection-host.directive';
import * as i0 from "@angular/core";
import * as i1 from "./selection-host.directive";
export class NgDocSelectionOriginDirective {
    constructor(elementRef, selectionHost) {
        this.elementRef = elementRef;
        this.selectionHost = selectionHost;
        this.selected = false;
        this.selectionHost.addOrigin(this);
    }
    ngOnChanges({ selected }) {
        if (selected) {
            this.selectionHost.changeSelected(this, this.selected);
        }
    }
    ngOnDestroy() {
        this.selectionHost.removeOrigin(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocSelectionOriginDirective, deps: [{ token: i0.ElementRef }, { token: i1.NgDocSelectionHostDirective }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.1.0", type: NgDocSelectionOriginDirective, isStandalone: true, selector: "[ngDocSelectionOrigin]", inputs: { selected: ["ngDocSelectionOrigin", "selected"] }, usesOnChanges: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocSelectionOriginDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngDocSelectionOrigin]',
                    standalone: true,
                }]
        }], ctorParameters: () => [{ type: i0.ElementRef }, { type: i1.NgDocSelectionHostDirective }], propDecorators: { selected: [{
                type: Input,
                args: ['ngDocSelectionOrigin']
            }] } });
