// @ng-doc/ui-kit/esm2022/observables/zone.mjs
import { Observable, pipe, } from 'rxjs';
class NgDocZoneDetachSourceOperator {
    constructor(ngZone) {
        this.ngZone = ngZone;
    }
    call(observer, source) {
        return this.ngZone.runOutsideAngular(() => source.subscribe(observer));
    }
}
/**
 *
 * @param ngZone
 */
function zoneDetachSource(ngZone) {
    return (source) => source.lift(new NgDocZoneDetachSourceOperator(ngZone));
}
/**
 *
 * @param ngZone
 */
function zoneDetach(ngZone) {
    return (source) => new Observable((subscriber) => source.subscribe({
        next: (value) => ngZone.runOutsideAngular(() => subscriber.next(value)),
        error: (error) => ngZone.runOutsideAngular(() => subscriber.error(error)),
        complete: () => ngZone.runOutsideAngular(() => subscriber.complete()),
    }));
}
/**
 *
 * @param ngZone
 */
export function ngDocZoneAttach(ngZone) {
    return (source) => new Observable((subscriber) => source.subscribe({
        next: (value) => ngZone.run(() => subscriber.next(value)),
        error: (error) => ngZone.run(() => subscriber.error(error)),
        complete: () => ngZone.run(() => subscriber.complete()),
    }));
}
/**
 *
 * @param ngZone
 */
export function ngDocZoneDetach(ngZone) {
    return pipe(zoneDetach(ngZone), zoneDetachSource(ngZone));
}
/**
 *
 * @param ngZone
 */
export function ngDocZoneOptimize(ngZone) {
    return pipe(zoneDetach(ngZone), zoneDetachSource(ngZone), ngDocZoneAttach(ngZone));
}
