// @ng-doc/ui-kit/fesm2022/ng-doc-ui-kit-directives-autofocus.mjs
import * as i0 from '@angular/core';
import { Directive, Input } from '@angular/core';
import { NgDocFocusUtils } from '@ng-doc/ui-kit/utils';

class NgDocAutofocusDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.selectAll = false;
    }
    ngOnInit() {
        const element = this.elementRef.nativeElement;
        if (NgDocFocusUtils.isNativeKeyboardFocusable(element)) {
            element.focus();
        }
        if (this.selectAll && element instanceof HTMLInputElement) {
            Promise.resolve().then(() => element.select());
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocAutofocusDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.1.0", type: NgDocAutofocusDirective, isStandalone: true, selector: "[ngDocAutofocus]", inputs: { selectAll: "selectAll" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocAutofocusDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngDocAutofocus]',
                    standalone: true,
                }]
        }], ctorParameters: () => [{ type: i0.ElementRef }], propDecorators: { selectAll: [{
                type: Input
            }] } });

/**
 * Generated bundle index. Do not edit.
 */

export { NgDocAutofocusDirective };
