// @ng-doc/ui-kit/fesm2022/ng-doc-ui-kit-directives-input-number.mjs
import { __decorate, __metadata } from 'tslib';
import * as i0 from '@angular/core';
import { forwardRef, Directive, HostListener } from '@angular/core';
import { isPresent } from '@ng-doc/core/helpers/is-present';
import { NgDocBaseInput } from '@ng-doc/ui-kit/classes/base-input';
import { toElement } from '@ng-doc/ui-kit/helpers';
import { UntilDestroy } from '@ngneat/until-destroy';

let NgDocInputNumberDirective = class NgDocInputNumberDirective extends NgDocBaseInput {
    constructor() {
        super({
            onIncomingUpdate: (value) => {
                toElement(this.elementRef).value = isPresent(value) ? String(Number(value)) : '';
            },
        });
    }
    blurEvent() {
        this.touch();
    }
    inputEvent() {
        this.updateModel(Number(this.elementRef.nativeElement.value));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocInputNumberDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.1.0", type: NgDocInputNumberDirective, isStandalone: true, selector: "input[ngDocInputNumber]", host: { listeners: { "blur": "blurEvent()", "input": "inputEvent()" } }, providers: [
            { provide: NgDocBaseInput, useExisting: forwardRef(() => NgDocInputNumberDirective) },
        ], usesInheritance: true, ngImport: i0 }); }
};
NgDocInputNumberDirective = __decorate([
    UntilDestroy(),
    __metadata("design:paramtypes", [])
], NgDocInputNumberDirective);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NgDocInputNumberDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: `input[ngDocInputNumber]`,
                    providers: [
                        { provide: NgDocBaseInput, useExisting: forwardRef(() => NgDocInputNumberDirective) },
                    ],
                    standalone: true,
                }]
        }], ctorParameters: () => [], propDecorators: { blurEvent: [{
                type: HostListener,
                args: ['blur']
            }], inputEvent: [{
                type: HostListener,
                args: ['input']
            }] } });

/**
 * Generated bundle index. Do not edit.
 */

export { NgDocInputNumberDirective };
