// @ng-doc/builder/helpers/entity-type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPageEntity = isPageEntity;
exports.isCategoryEntity = isCategoryEntity;
exports.isSourceFileEntity = isSourceFileEntity;
exports.isNavigationEntity = isNavigationEntity;
exports.isApiEntity = isApiEntity;
exports.isFileEntity = isFileEntity;
exports.isApiPageEntity = isApiPageEntity;
exports.isApiScopeEntity = isApiScopeEntity;
exports.isRouteEntity = isRouteEntity;
const engine_1 = require("../engine");
const file_entity_1 = require("../engine/entities/abstractions/file.entity");
const navigation_entity_1 = require("../engine/entities/abstractions/navigation.entity");
const route_entity_1 = require("../engine/entities/abstractions/route.entity");
const source_file_entity_1 = require("../engine/entities/abstractions/source-file.entity");
/**
 *
 * @param page
 */
function isPageEntity(page) {
    return page instanceof engine_1.NgDocPageEntity;
}
/**
 *
 * @param page
 */
function isCategoryEntity(page) {
    return page instanceof engine_1.NgDocCategoryEntity;
}
/**
 *
 * @param page
 */
function isSourceFileEntity(page) {
    return page instanceof source_file_entity_1.NgDocSourceFileEntity;
}
/**
 *
 * @param page
 */
function isNavigationEntity(page) {
    return page instanceof navigation_entity_1.NgDocNavigationEntity;
}
/**
 *
 * @param page
 */
function isApiEntity(page) {
    return page instanceof engine_1.NgDocApiEntity;
}
/**
 *
 * @param page
 */
function isFileEntity(page) {
    return page instanceof file_entity_1.NgDocFileEntity;
}
/**
 *
 * @param page
 */
function isApiPageEntity(page) {
    return page instanceof engine_1.NgDocApiPageEntity;
}
/**
 *
 * @param page
 */
function isApiScopeEntity(page) {
    return page instanceof engine_1.NgDocApiScopeEntity;
}
/**
 *
 * @param page
 */
function isRouteEntity(page) {
    return page instanceof route_entity_1.NgDocRouteEntity;
}
